### ISSUE TITLE
Multi Workspace / Multi Slice Viewport Per Scene

### ISSUE THREAD

**Issue by ZeHipo** [godotengine/godot-proposals#2570]
~ sections A through D correspond to the issue template ~

A::
game: working on a 3d touchscreen game.

B::
problems encountered: OCD (not seeing code and game at the same time), problem solving (visual aid while scripting), editor workflow (disconnect between current script and previously selected scene in 2d workspace). 

having the 2d workspace and script workspace open at the same time, side by side, helps to think about the code for a gui (layout etc), having 3d and 2d workspace open at the same time helps with the layout of eg touch controls.

C::
proposal: divide the viewport in "slices" each with different workspaces.

my proposal would be putting the current workspace selector into the scene tab like there already is a script symbol in the tab.

right clicking on eg the script symbol in the scene tab opens the script workspace in a new "viewport slice", splitting the current viewport (eg 2d) in half, with its own toolbar and bottom panel. this layout would be saved per scene, so you could work on one scene in double or triple view and work on another one in single view.

when the additional view is not needed in can be closed from its toolbar with an "X".

the dividing line between the viewport slices can be dragged to change the size ratio of the two workspace views.

right clicking the dividing line opens a pop-up prompt at mouse location which allows the neighboring slices to swap sides.

simply changing workspaces would be done by left clicking on icons in the scene tab instead of the current global workspace selector. to save space it would be a drop down menu.

i'm thinking about vertical slices (columns), but horizontal slices could be an option. after selecting what new view should be opened (with right click) a dotted line across the viewport follows the mouse. when mousing over the top or bottom edge the line is vertical, when mousing over the left or right side it is horizontal. left clicking confirms the size and orientation of the new slice.

D:: 
complexity / lines of code and frequency of use, possible implementation as an add-on:
i understand that this is a major change to the Godot editor as a whole and probably requires its own branch on the level of a new release. i'm the first one to admit that this level of coding is way out of my league.

E::
multi window: a multi window workflow is planned for future Godot releases. this proposal of viewport slices might help keeping a multi window setup tidy and manageable by allowing to use one window for one scene with all information accessible at a glance.

F::
in large, industry standard teams there seems to be a sharp division between artist of different domains and coders. in smaller indie teams designers work on the game more interconnectedly / as a whole / wear many hats during development. in this latter case it is advantageous to have the game visible from multiple contexts simultaneously.

G::
step by step:
you have a scene open in 2d view. the scene tab looks like this from left to right: class icon, name of the scene, icon of the current workspace, in this case it is "2d". 

( there is no global workspace selector any more. since the AssetLib is planned for overhaul it's placement is not part of this proposal. )

left clicking on the "2d" icon opens a drop down menu under the mouse with icons for "3d" and "script" icons under the "2d". mousing away closes the drop down after a short while. clicking on one of the other icons changes the viewport like now (Godot 3).

RIGHT clicking on the same icon also opens this drop down, but this time the icons have a "+" sign next to them, like a badge on a messaging app on a smart phone. clicking on a "3d+" or "script+" icon closes the drop down and and triggers a dotted line horizontally or vertically across the viewport which follows the mouse.

the icons of the future workspace slices appear next to the dotted line. left and right of the vertical line (at the top of the viewport) or above and below the horizontal line (left edge of the window). by default the new workspace takes the right or bottom slice.

left click confirms the placement of the dividing line between the new slices and adds the new workspace slice to the viewport. 
right click swaps the sides of the two future slices but doesn't confirm (the icons next to the line swap sides). 
escape cancels the action.

after left clicking the current viewport is divided in two and the new viewport takes one of the halves. each slice it has its own toolbar and bottom panel. additionally a "X" icon appears in each toolbar.

when one of the slices is not needed any more clicking on its toolbar "X" icon closes it and the slice left to it expands over its screen space (unless it is the left most slice).

**Comment by Calinou** [comment]
@ZeHipo Could you follow the [issue template](https://github.com/godotengine/godot-proposals/blob/master/.github/ISSUE_TEMPLATE/feature---enhancement-proposal.md) and add headings above each section?

**Comment by Calinou** [comment]
In the interest of making high-quality proposals, please [read the rules](https://github.com/godotengine/godot-proposals#readme) and fill in the issue template. This proposal will be reopened once this is done :slightly_smiling_face:
